### PR TITLE
935 fee token l1 contract

### PR DESCRIPTION
**Pull request by alysiahuggins**
- created the fee contract
- made it upgradeable
- added tests

### CODE REVIEW

**Comment by philippecamacho** [comment]
@alysiahuggins I made some minor changes:
 * 687d1e0b9329de01ee1288bdd16d911fe26f2601 
   This strange value for `ETHERSCAN_API_KEY` is needed to avoid some annoying warnings. I added a [comment](47a9565121127966cdf3396c34e42799dc444ce3).
   * ac55d525b55ecdfb1c7b462e56bf520941ad8992
   By default the contracts which name contain "Test" are omitted when generating bindings.
   
   Now a question: do we need the bindings for the new contracts like `uups_upgradeable.rs`?

File: contracts/src/FeeContract.sol
Comment (philippecamacho) [inline contracts/src/FeeContract.sol:70]: This if-then statement is redundant with the next one, maybe we can remove it?

File: contracts/src/FeeContract.sol
Comment (philippecamacho) [inline contracts/src/FeeContract.sol:82]: I think we do not need to track the balances of each user in the contract as the accounting will happen inside HotShot. Is the `balances` mapping for testing purposes only?

File: contracts/src/FeeContract.sol
Comment (alysiahuggins) [reply contracts/src/FeeContract.sol:70]: ah yes, very true, i had this if statement before I added the other

Comment (alysiahuggins) [reply contracts/src/FeeContract.sol]: fix: https://github.com/EspressoSystems/espresso-sequencer/commit/fa35ff3bdb59e938fc0895b3dd6b7d4d0cba5e64

File: contracts/src/FeeContract.sol
Comment (alysiahuggins) [reply contracts/src/FeeContract.sol:78]: Oh I see, but would hotshot need to read the balance deposited from this contract initially? Also when the user deposits more

File: contracts/script/DeployFeeContract.s.sol
Comment (alxiong) [inline contracts/script/DeployFeeContract.s.sol:31]: I actually didn't fully understand how this works. 
[`ERC1967Proxy.constructor`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/ERC1967/ERC1967Proxy.sol#L27) will call `ERC1967Utils.upgradeToAndCall` which [would call ](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/ERC1967/ERC1967Utils.sol#L86)`Address.functionDelegateCall`, which [would call ](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Address.sol#L105)the constructor on fee contract with `msg.sender`

but our fee contract constructor doesn't accept any input argument

what did I miss?

File: contracts/test/FeeContract.t.sol
Comment (alxiong) [inline contracts/test/FeeContract.t.sol:169]: if we set the `admin` to what's derived from env.mnemonic, then we wouldn't need this deployer contract variable, just storing the `address proxy` is sufficient.

File: contracts/test/FeeContract.t.sol
Comment (alxiong) [inline contracts/test/FeeContract.t.sol:254]: we should add `expectRevert` ?

also should we test that if the we can upgrade to a `V2` with no admin? and/or become non-upgradable?

**Review by alxiong — COMMENTED** [review]
👍 good job!

- when expected revert, we should also try to match the thrown error type to make sure we are catching the expected revert, not accidentally caused by other reasons.
- remember to remove commented out code (such as `//console.log`, or some lines in deployment scripts

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol:86]: yea good point, I'll update these

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol:28]: yea that makes sense, i'll remove testFuzz_depositTwice(), replace with depositMany which involves a for loop. good catch

File: contracts/script/DeployFeeContract.s.sol
Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol:13]: well actually when I deployed to testnet, i passed the private key in as an env variable, however, would we use a multisig for the actual deployment?

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol]: https://github.com/EspressoSystems/espresso-sequencer/commit/132ced07ad629b957f13633a42a7937cd337230b

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol]: https://github.com/EspressoSystems/espresso-sequencer/commit/428d568966fadfb87338b9535d0430b88df40419

File: contracts/script/DeployFeeContract.s.sol
Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: https://github.com/EspressoSystems/espresso-sequencer/commit/ca547ffc8dbcab42a8830ab3f5e8b193d6d6864d

File: contracts/script/DeployFeeContract.s.sol
Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: https://github.com/EspressoSystems/espresso-sequencer/commit/ca547ffc8dbcab42a8830ab3f5e8b193d6d6864d

File: contracts/script/DeployFeeContract.s.sol
Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: > why separate function, should inline it

https://github.com/EspressoSystems/espresso-sequencer/commit/fd5bdd1b09d29146c1df649d7a64d100ba5e783f

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol]: true, there is no need for the deployer to be a state variable https://github.com/EspressoSystems/espresso-sequencer/commit/3c7c6fe7fd90130907e29bd90863354a7cd58bea

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol:241]: > also should we test that if the we can upgrade to a V2 with no admin? and/or become non-upgradable?
so test whether it's possible to upgrade to another contract that isn't upgradeable?

Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol:241]: > we should add expectRevert ?
I used the `testFail` syntax which means that the foundry test fails when the contract call reverts. Currently checking with the team to see if there is a preferred testing method for expected failures.

Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol]: > > we should add expectRevert ?
> 
> I used the `testFail` syntax which means that the foundry test fails when the contract call reverts. Currently checking with the team to see if there is a preferred testing method for expected failures.

since we decided to go with expectRevert, I've updated the test https://github.com/EspressoSystems/espresso-sequencer/commit/d4e477df94a8b75a1dd3ec1a30d75363afc5fb4d

File: contracts/src/FeeContract.sol
Comment (alysiahuggins) [reply contracts/src/FeeContract.sol:78]: any thoughts on this @alxiong @jbearer ?

File: contracts/script/DeployFeeContract.s.sol
Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: Yea the FeeContract's constructor is actually disabled, so the function that's actually being called is the `initialize` method on `FeeContract.sol`

File: contracts/src/FeeContract.sol
Comment (philippecamacho) [reply contracts/src/FeeContract.sol:78]: While it is possible, I don't think it is needed as it should be enough to read from the deposit events. That said it depends on the implementation of the fee accounting inside HotShot cc @jbearer.

Comment (philippecamacho) [reply contracts/src/FeeContract.sol:78]: Anyways, this function cannot create any harm and makes testing simpler so I am ok to keep it.

Comment (alysiahuggins) [reply contracts/src/FeeContract.sol:78]: very true @ being able to read from the deposit events. Right now, we assume that Hotshot takes care of all the accounting and also storage. So maybe it's worth considering removing it so that the L1 smart contract storage doesn't increase which can impact gas fees over time. Unless we want to consider things like onchain (Ethereum L1) integrity and a possible future where we handle ERC20 tokens and then give the users a wrapped version that's represented in their wallet?

Comment (alxiong) [reply contracts/src/FeeContract.sol:78]: personally, i prefer to keep the map in contract, it's not _very_ expensive as it's minimally simple.

having this map makes upgrade with "less" trust, instead of completely at the arbitration of the admin (i'd imagine ppl trusting Espresso, but not when there's many millions deposited and our admin and just "oops, I upgrade a contract that can claim a different mapping, users can prove historical event to claim their coin if they are so keen")

File: contracts/script/DeployFeeContract.s.sol
Comment (alxiong) [reply contracts/script/DeployFeeContract.s.sol]: if we are using multisig, we should also be loading the multisig account/address from env var, right?

if so, here loading from `envString` is closer to actual code?

Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: So it depends on how we manage it. for example, if we use Openzeppelin Defender for managing the deployment of the contracts the script would look like this https://github.com/OpenZeppelin/openzeppelin-foundry-upgrades/blob/main/test/Defender.s.sol (though at the moment this is not sufficient since deployments with defender via Foundry doesn't support constructor arguments, but I think Hardhat + Defender does). So I can uncomment this code for now but it's likely to change.

File: contracts/script/DeployFeeContract.s.sol
Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: oh yeaaa, good point, it just uses the msg.sender that calls the method, so I can comment that code out

Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: https://github.com/EspressoSystems/espresso-sequencer/commit/7df709c6ea1e027957ebd43e507e080eb4a7e2d0

**Review by alxiong — COMMENTED** [review]
✅ 
The code looks good, but pls wait until #997 is merged, then sync this with main, as they would exclude all the unnecessary bindings generated. 

btw, should we also remove the `broadcasts/*` artifacts? @alysiahuggins

File: contracts/script/DeployFeeContract.s.sol
Comment (alxiong) [reply contracts/script/DeployFeeContract.s.sol]: thanks for the info! we can leave it as is for now

Comment (alysiahuggins) [reply contracts/script/DeployFeeContract.s.sol]: https://github.com/EspressoSystems/espresso-sequencer/commit/65917037d9433e95af5fe4c54c7060535b7ebe74

**Comment by alysiahuggins** [comment]
> ✅ The code looks good, but pls wait until #997 is merged, then sync this with main, as they would exclude all the unnecessary bindings generated.
> 
> btw, should we also remove the `broadcasts/*` artifacts? @alysiahuggins

Hey, let me know if all is ready to merge especially since we excluded the forge bindings issue from #997 and put it into a new issue, #1068. As for the broadcast folder, are you sure you want to remove it? It was a real deployment to sepolia testnet?

**Comment by alysiahuggins** [comment]
> > ✅ The code looks good, but pls wait until #997 is merged, then sync this with main, as they would exclude all the unnecessary bindings generated.
> > btw, should we also remove the `broadcasts/*` artifacts? @alysiahuggins
> 
> Hey, let me know if all is ready to merge especially since we excluded the forge bindings issue from #997 and put it into a new issue, #1068. As for the broadcast folder, are you sure you want to remove it? It was a real deployment to sepolia testnet?

sepolia deployment broadcast logs removed since it's not a part of the public testnet release https://github.com/EspressoSystems/espresso-sequencer/commit/2e193c09908c2a4f02bccff8218228656cfaaf9d

File: contracts/test/FeeContract.t.sol
Comment (alxiong) [inline contracts/test/FeeContract.t.sol:87]: i think this will always be `MIN_DEPOSIT_AMOUNT` as the integer division will always round to zero?

**Review by alxiong — APPROVED** [review]
LGTM

File: contracts/test/FeeContract.t.sol
Comment (alysiahuggins) [reply contracts/test/FeeContract.t.sol]: oh that's right!! the test still proves what it sets out to prove but I'll remove the division to make it easier https://github.com/EspressoSystems/espresso-sequencer/commit/b1884c4bea3246a86cbd430853a5a69e4def0f0a

**Review by philippecamacho — APPROVED** [review]
LGTM

**Pull request merged**
